src/sagemaker/git_utils.py

```diff
@@ -72,44 +81,182 @@ def git_clone_repo(git_config, entry_point, source_dir=None, dependencies=None):
+    updated_url = git_config["repo"]
+    if exists["token"]:
+        if exists["username"] or exists["password"]:
+            warnings.warn(
+                "Using token for authentication, "
+                "but unnecessary credential argument(s) provided."
+            )
+        updated_url = _insert_token_to_repo_url(url=git_config["repo"], token=git_config["token"])
+    elif exists["username"] and exists["password"]:
+        updated_url = _insert_username_and_password_to_repo_url(
+            url=git_config["repo"], username=git_config["username"], password=git_config["password"]
+        )
+    elif exists["username"] or exists["password"]:
+        warnings.warn("Unnecessary credential argument(s) provided.")
+    _run_clone_command(updated_url, repo_dir)
+
+
+def _clone_command_for_github_like_https_2fa_enabled(git_config, repo_dir, exists):
+    updated_url = git_config["repo"]
+    if exists["token"]:
+        if exists["username"] or exists["password"]:
+            warnings.warn(
+                "Using token for authentication, "
+                "but unnecessary credential argument(s) provided."
+            )
+        updated_url = _insert_token_to_repo_url(url=git_config["repo"], token=git_config["token"])
+    elif exists["username"] or exists["password"] or exists["token"]:
+        warnings.warn(
+            "Unnecessary credential argument(s) provided."
+            "Hint: since two factor authentication is enabled, you have to provide token."
+        )
+    _run_clone_command(updated_url, repo_dir)
+
+
+def _run_clone_command(repo_url, repo_dir):
+    """Run the 'git clone' command with the repo url and the directory to clone the repo into.
+
+    Args:
+        repo_url (str): Git repo url to be cloned.
+        repo_dir: (str): Local path where the repo should be cloned into.
+
+    Raises:
+        CalledProcessError: If failed to clone git repo.
+    """
+    my_env = os.environ.copy()
+    if repo_url.startswith("https://"):
+        my_env["GIT_TERMINAL_PROMPT"] = "0"
+    elif repo_url.startswith("git@"):
+        f = tempfile.NamedTemporaryFile()
+        w = open(f.name, "w")
+        w.write("ssh -oBatchMode=yes $@")
+        w.close()
+        # 511 in decimal is same as 777 in octal
+        os.chmod(f.name, 511)
+        my_env["GIT_SSH"] = f.name
+    subprocess.check_call(["git", "clone", repo_url, repo_dir], env=my_env)
+
+
+def _insert_token_to_repo_url(url, token):
+    """Insert the token to the Git repo url, to make a component of the git clone command. This method can
+    only be called when repo_url is an https url.
+
+    Args:
+        url (str): Git repo url where the token should be inserted into.
+        token (str): Token to be inserted.
+
+    Returns:
+        str: the component needed fot the git clone command.
+    """
+    index = len("https://")
+    return url[:index] + token + "@" + url[index:]
```

Comment:
tip: use replace

```python


url.replace('https://', 'https://' + token + '@')
```

Reply:
We need to handle the case which the token and ```@``` are already included in the url.

Reply:
Have dealt with that situation. Do we have the same concern for username and password? I am thinking maybe not because people usually won't first encode the password and then insert it into url.